pageserver/src/tenant.rs

```diff
@@ -639,26 +641,62 @@ impl Tenant {
             .as_ref()
             .ok_or_else(|| anyhow::anyhow!("cannot attach without remote storage"))?;
 
-        let remote_timelines = remote_timeline_client::list_remote_timelines(
+        let remote_timeline_ids = remote_timeline_client::list_remote_timelines(
             remote_storage,
             self.conf,
             self.tenant_id,
         )
         .await?;
 
-        info!("found {} timelines", remote_timelines.len());
+        info!("found {} timelines", remote_timeline_ids.len());
 
-        let mut timeline_ancestors: HashMap<TimelineId, TimelineMetadata> = HashMap::new();
-        let mut index_parts: HashMap<TimelineId, IndexPart> = HashMap::new();
-        for (timeline_id, index_part) in remote_timelines {
-            let remote_metadata = index_part.parse_metadata().with_context(|| {
-                format!(
-                    "Failed to parse metadata file from remote storage for tenant {} timeline {}",
-                    self.tenant_id, timeline_id
-                )
-            })?;
+        // Download & parse index parts
+        let mut part_downloads = JoinSet::new();
+        for timeline_id in remote_timeline_ids {
+            let client = RemoteTimelineClient::new(
+                remote_storage.clone(),
+                self.conf,
```

Comment:
why not pass Tenant (self) and let RemoteTimelineClient take what it needs ?

Reply:
`RemoteTimelineClient` doesn't need anything from `Tenant` except the `TenantId` and `TimelineId`.
So, limit its view to what matters.
I guess some people call this "Interface Segregation Principle".

Also, it makes `RemoteTimelineClient`  easier to test because we don't have to construct a `Tenant` object for it.

Finally, in Rust, if we passed the `Arc<Tenant>`, or better `Arc<Timeline>` down into `RemoteTimelineClient`, we'd have a reference cycle: `Timeline => RemoteTimelineClient => Timeline`